powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/EciesUtils.java

```diff
@@ -62,27 +62,50 @@ public static long generateTimestamp() {
      * @param protocolVersion Protocol version.
      * @param applicationKey Application key.
      * @param activationId Activation ID.
+     * @param temporaryKeyId Temporary key ID.
      * @return Derived associated data.
      * @throws EciesException In case that activation ID is required but is missing.
      */
-    public static byte[] deriveAssociatedData(EncryptorScope scope, String protocolVersion, String applicationKey, String activationId) throws EciesException {
+    public static byte[] deriveAssociatedData(EncryptorScope scope, String protocolVersion, String applicationKey, String activationId, String temporaryKeyId) throws EciesException {
         if (protocolVersion == null) {
             throw new EciesException("Protocol version is missing");
         }
-        if ("3.2".equals(protocolVersion)) {
-            if (applicationKey == null) {
-                throw new EciesException("Application key is missing");
+        switch (protocolVersion) {
+            case "3.2": {
+                if (applicationKey == null) {
+                    throw new EciesException("Application key is missing");
+                }
+                if (scope == EncryptorScope.ACTIVATION_SCOPE) {
+                    if (activationId == null) {
+                        throw new EciesException("Activation ID is missing in ACTIVATION_SCOPE");
+                    }
+                    return ByteUtils.concatStrings(protocolVersion, applicationKey, activationId);
+                } else {
+                    return ByteUtils.concatStrings(protocolVersion, applicationKey);
+                }
             }
-            if (scope == EncryptorScope.ACTIVATION_SCOPE) {
-                if (activationId == null) {
-                    throw new EciesException("Activation ID is missing in ACTIVATION_SCOPE");
+            case "3.3": {
+                if (applicationKey == null) {
+                    throw new EciesException("Application key is missing");
+                }
+                if (scope == EncryptorScope.ACTIVATION_SCOPE) {
+                    if (activationId == null) {
+                        throw new EciesException("Activation ID is missing in ACTIVATION_SCOPE");
+                    }
+                    if (temporaryKeyId == null) {
+                        throw new EciesException("Missing temporary key identifier");
+                    }
```

Comment:
You can test this for both scopes at once, just move the test before condition: 
```java
if (scope == EncryptorScope.ACTIVATION_SCOPE) {
```

Reply:
Good idea...